### PR TITLE
Some clippy fixes in the compiler

### PR DESCRIPTION
**Pull request by Noratrieb**
Best reviewed commit-by-commit 📎.

### CODE REVIEW

File: compiler/rustc_session/src/parse.rs [hunk 1]
```diff
@@ -98,20 +98,21 @@ pub fn feature_err<'a>(
...
-) -> DiagnosticBuilder<'a, ErrorGuaranteed> {
+) -> DiagnosticBuilder<'_, ErrorGuaranteed> {
     let span = span.into();
 
     // Cancel an earlier warning for this same error, if it exists.
     if let Some(span) = span.primary_span() {
-        sess.span_diagnostic
-            .steal_diagnostic(span, StashKey::EarlySyntaxWarning)
-            .map(|err| err.cancel());
+        if let Some(err) = sess.span_diagnostic.steal_diagnostic(span, StashKey::EarlySyntaxWarning)
```
Comment (compiler-errors) [inline compiler/rustc_session/src/parse.rs:112]: good

File: compiler/rustc_resolve/src/late.rs [hunk 2]
```diff
@@ -1224,7 +1224,7 @@ impl<'a: 'ast, 'b, 'ast, 'tcx> LateResolutionVisitor<'a, 'b, 'ast, 'tcx> {
             lifetime_ribs: Vec::new(),
             lifetime_elision_candidates: None,
             current_trait_ref: None,
-            diagnostic_metadata: Box::new(DiagnosticMetadata::default()),
+            diagnostic_metadata: Box::<DiagnosticMetadata<'_>>::default(),
```
Comment (compiler-errors) [inline compiler/rustc_resolve/src/late.rs:1227]: i dislike this lol

File: compiler/rustc_arena/src/lib.rs [hunk 3]
```diff
@@ -22,6 +22,7 @@
 #![feature(strict_provenance)]
 #![deny(rustc::untranslatable_diagnostic)]
 #![deny(rustc::diagnostic_outside_of_impl)]
+#![allow(clippy::mut_from_ref)] // Arena allocators are one of the places where this pattern is fine.
```
Comment (saethlin) [inline compiler/rustc_arena/src/lib.rs:25]: Isn't this redundant with the other outer allows in this file?

Comment (Nilstrieb) [reply compiler/rustc_arena/src/lib.rs:25]: Lint levels aren't hygienic, so the macro sadly still needs them.

File: compiler/rustc_middle/src/traits/mod.rs [hunk 4]
```diff
@@ -923,7 +923,7 @@ impl ObjectSafetyViolation {
                 }
             }
             ObjectSafetyViolation::SupertraitNonLifetimeBinder(_) => {
-                format!("where clause cannot reference non-lifetime `for<...>` variables").into()
```
Comment (compiler-errors) [inline compiler/rustc_middle/src/traits/mod.rs:926]: who wrote this :forgor:

File: compiler/rustc_middle/src/mir/interpret/allocation.rs [hunk 5]
```diff
@@ -109,26 +109,34 @@ const MAX_HASHED_BUFFER_LEN: usize = 2 * MAX_BYTES_TO_HASH;
 // large.
 impl hash::Hash for Allocation {
     fn hash<H: hash::Hasher>(&self, state: &mut H) {
+        let Self {
+            bytes,
+            provenance,
+            init_mask,
+            align,
+            mutability,
+            extra: _, // don't bother hashing ()
```
Comment (saethlin) [inline compiler/rustc_middle/src/mir/interpret/allocation.rs:118]: Extra is not always `()`, so I think it is wrong to skip hashing the field altogether.

File: compiler/rustc_middle/src/mir/interpret/allocation.rs [hunk 6]
```diff
@@ -109,26 +109,34 @@ const MAX_HASHED_BUFFER_LEN: usize = 2 * MAX_BYTES_TO_HASH;
 // large.
 impl hash::Hash for Allocation {
     fn hash<H: hash::Hasher>(&self, state: &mut H) {
+        let Self {
```
Comment (compiler-errors) [inline compiler/rustc_middle/src/mir/interpret/allocation.rs:112]: is this really a clippy lint

File: compiler/rustc_macros/src/diagnostics/fluent.rs [hunk 7]
```diff
@@ -100,7 +100,7 @@ pub(crate) fn fluent_messages(input: proc_macro::TokenStream) -> proc_macro::Tok
             Diagnostic::spanned(
                 resource_span,
                 Level::Error,
-                format!("could not open Fluent resource: {}", e.to_string()),
+                format!("could not open Fluent resource: {}", e),
```
Comment (compiler-errors) [inline compiler/rustc_macros/src/diagnostics/fluent.rs:103]:
```suggestion
                format!("could not open Fluent resource: {e}"),
```

File: compiler/rustc_middle/src/mir/interpret/allocation.rs (hunk 5 above)
Comment (Nilstrieb) [reply compiler/rustc_middle/src/mir/interpret/allocation.rs:118]: Inside the impl block it is

File: compiler/rustc_data_structures/src/profiling.rs [hunk 8]
```diff
@@ -778,7 +778,7 @@ pub fn print_time_passes_entry(
                 "rss_start": start_rss,
                 "rss_end": end_rss,
             });
-            eprintln!("time: {}", json.to_string());
+            eprintln!("time: {}", json);
```
Comment (compiler-errors) [inline compiler/rustc_data_structures/src/profiling.rs:781]:
```suggestion
            eprintln!("time: {json}");
```

?

File: compiler/rustc_arena/src/lib.rs (hunk 3 above)
Comment (saethlin) [reply compiler/rustc_arena/src/lib.rs:25]: Ew.

**Review by compiler-errors — COMMENTED** [review]
r=me

File: compiler/rustc_middle/src/mir/interpret/allocation.rs (hunk 5 above)
Comment (saethlin) [reply compiler/rustc_middle/src/mir/interpret/allocation.rs:118]: I'm also not sure it's worth adding any complexity to not hash `()`, does that completely optimize out?

Comment (compiler-errors) [reply compiler/rustc_middle/src/mir/interpret/allocation.rs:118]: can you match it like `extra: ()`?

File: compiler/rustc_middle/src/mir/interpret/allocation.rs (hunk 6 above)
Comment (Nilstrieb) [reply compiler/rustc_middle/src/mir/interpret/allocation.rs:112]: it's a nils lint

**Comment by Nilstrieb** [comment]
I removed the `Don't assume that `io::{Read, Write}` writes/reads all` commit since that seems to be causing an ICE that I can't be bothered to further investigate, the code is fine as-is even though it technically relies on unspecified details of `Read,Write`.

File: compiler/rustc_middle/src/mir/interpret/allocation.rs (hunk 5 above)
Comment (Nilstrieb) [reply compiler/rustc_middle/src/mir/interpret/allocation.rs:118]: great idea, I will.

File: compiler/rustc_resolve/src/late.rs (hunk 2 above)
Comment (Nilstrieb) [reply compiler/rustc_resolve/src/late.rs:1227]: I can remove the turbofish if you want^^

**Comment by saethlin** [comment]
How do we repro the ICE? Seems worth someone writing an issue

File: compiler/rustc_middle/src/traits/mod.rs (hunk 4 above)
Comment (Nilstrieb) [reply compiler/rustc_middle/src/traits/mod.rs:926]: git blame now points at me, so it's clearly my fault

File: compiler/rustc_resolve/src/late.rs (hunk 2 above)
Comment (compiler-errors) [reply compiler/rustc_resolve/src/late.rs:1227]: Or maybe just `Default::default()` lol

**Comment by Nilstrieb** [comment]
I used `read_exact` but there aren't enough bytes in the data to fill the buffer, returning an error. `read`  just returned as many bytes as there are, which is the desired behavior here. The code isn't wrong so it's fine to leave it, but it should maybe be rewritten in the future to not do it like that, not that it's very important.
a90bd02df386787c621cdfcc17b96a8e3ecb13e9

**Comment by compiler-errors** [comment]
Ok, looking at a90bd02, I *think* it makes sense that the `read_all` would fail -- we're trying to read at most a u128 from the source buffer, but not *necessarily* a u128, so source is not necessarily at least sizeof(u128) long. Anyways, looks fine now.

r? @compiler-errors @bors r+

File: compiler/rustc_data_structures/src/stack.rs [hunk 9]
```diff
@@ -5,7 +5,7 @@ const RED_ZONE: usize = 100 * 1024; // 100k
 
 // Only the first stack that is pushed, grows exponentially (2^n * STACK_PER_RECURSION) from then
 // on. This flag has performance relevant characteristics. Don't set it too high.
-const STACK_PER_RECURSION: usize = 1 * 1024 * 1024; // 1MB
+const STACK_PER_RECURSION: usize = 1024 * 1024; // 1MB
```
Comment (scottmcm) [inline compiler/rustc_data_structures/src/stack.rs:8]: TBH, I think clippy is wrong to complain about this.  Especially in a `const`, there's absolutely nothing wrong with a `1 *`, and representing "X MB" as `X * 1024 * 1024` is *fine*.

Maybe use `1 << 20`?  That's also clearly `1 MiB`, and hopefully clippy wouldn't whine about it.

Comment (saethlin) [reply compiler/rustc_data_structures/src/stack.rs:8]: I wish I agreed `1 << 20` was obviously 1 MB. Maybe I'm just uncommonly bad at all these bit twiddling operations.

Comment (scottmcm) [reply compiler/rustc_data_structures/src/stack.rs:8]: (For context: `k` is a *decimal* shift by 3, `M` is a decimal shift by 6, etc.  The elementary school "move the decimal point" is, essentially, a shift.  So `M` ≡ 10³, aka "<<₁₀ 3".  And similarly `Mi` is 2²⁰, aka `<< 20`.  So "kebi", "mebi", "gibi", "tebi", etc are "<< 10", "<< 20", "<< 30", "<< 40", etc.)

**Pull request merged**